executor/analyze.go

```diff
@@ -785,15 +949,37 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
 		topns = append(topns, topn)
 		fmSketches = append(fmSketches, rootRowCollector.FMSketches[i])
 	}
-	colLen := len(e.colsInfo)
+
+	indexPushedDownResult := <-idxNDVPushDownCh
+	if indexPushedDownResult.err != nil {
+		return 0, nil, nil, nil, nil, indexPushedDownResult.err
+	}
+	for _, offset := range indexesWithVirtualColOffsets {
+		ret := indexPushedDownResult.results[e.indexes[offset].ID]
+		rootRowCollector.NullCount[colLen+offset] = ret.Count
+		rootRowCollector.FMSketches[colLen+offset] = ret.Fms[0]
+	}
+
+	// build index stats
+	var tmpDatum types.Datum
 	for i, idx := range e.indexes {
 		sampleItems := make([]*statistics.SampleItem, 0, rootRowCollector.MaxSampleSize)
 		for _, row := range rootRowCollector.Samples {
 			if len(idx.Columns) == 1 && row.Columns[idx.Columns[0].Offset].IsNull() {
 				continue
 			}
+
 			b := make([]byte, 0, 8)
 			for _, col := range idx.Columns {
+				if col.Length != types.UnspecifiedLength {
+					row.Columns[col.Offset].Copy(&tmpDatum)
+					ranger.CutDatumByPrefixLen(&tmpDatum, col.Length, &e.colsInfo[col.Offset].FieldType)
+					b, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, b, tmpDatum)
+					if err != nil {
+						return 0, nil, nil, nil, nil, err
+					}
+					continue
+				}
 				b, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, b, row.Columns[col.Offset])
```

Comment:
What if we failed to decode virtual columns and this column is virtual? Will it cause a panic?